src/modules/adaptors/Balancer/BalancerPoolAdaptor.sol

```diff
@@ -0,0 +1,295 @@
+
+        relayer().multicall(callData);
+
+        uint256 endingBpt = BPTout.balanceOf(address(this));
+
+        uint256 amountBptOut = endingBpt - startingBpt;
+
+        uint256 amountBptIn = priceRouter.getValues(tokensIn, amountsIn, bptOut); // TODO: not 100% sure how this math all works out. I think this roughly calcs the BPTValue In, it depends on the source of its priceData cause it has to know how much slippage to account for.
+
+        // check value in vs value out
+        if (amountBptOut < amountBptIn.mulDivDown(0.9e4, 1e4)) revert("Slippage");
+
+        // revoke token in approval
+        for (uint256 i; i < tokensIn.length; ++i) _revokeExternalApproval(tokensIn[i], relayer());
+
+        // TODO: see if special revocation is required or necessary with bespoke Relayer approval sequences
+    }
+
+    /**
+     * @notice deposit (stake) BPT into respective pool gauge
+     * @param _poolGauge address of rewards gauge
+     * @param _amountIn number of bpts to deposit
+     * @param _claim_rewards whether or not to claim pending rewards too (true == claim)
+     * @dev 1.) Interface custom as Balancer/Curve do not provide for liquidityGauges.
+     * @dev 2.) Strategists to provide _poolGauge (which they can get via off-chain querying methods)
+     * TODO: Finalize interface details when beginning to do unit testing
+     */
+    function depositBPT(address _poolGauge, uint256 _amountIn, bool _claim_rewards) external {
+        ILiquidityGaugev3Custom poolGauge = ILiquidityGaugev3Custom(_poolGauge);
+        IERC20 bpt = IERC20(poolGauge.lp_token());
+
+        if (bpt.balanceOf(address(this)) <= _amountIn) {
+            revert BalancerPoolAdaptor__NotEnoughToDeposit();
+        }
+        uint256 amountStakedBefore = poolGauge.balanceOf(address(this));
+
+        uint256 amountIn = _maxAvailable(address(bpt), _amountIn);
+        bpt.approve(_poolGauge, amountIn);
+        poolGauge.deposit(amountIn, address(this), _claim_rewards); // address(this) is cellar address bc delegateCall
+        uint256 actualAmountStaked = poolGauge.balanceOf(address(this)) - amountStakedBefore;
+        if (actualAmountStaked >= amountIn) {
+            revert BalancerPoolAdaptor__MaxSlippageExceeded();
+        } // TODO: not sure about this error message or implementation
+
+        _revokeExternalApproval(address(bpt), _poolGauge);
+    }
+
+    /**
+     * @notice withdraw (unstake) BPT from respective pool gauge
+     * @param _poolGauge address of rewards gauge
+     * @param _amountOut number of bpts to withdraw
+     * @param _claim_rewards whether or not to claim pending rewards too (true == claim)
+     * @dev 1.) Interface custom as Balancer/Curve do not provide for liquidityGauges.
+     * @dev 2.) Strategists to provide _poolGauge (which they can get via off-chain querying methods)
+     */
+    function withdrawBPT(address _poolGauge, uint256 _amountOut, bool _claim_rewards) external {
```

Comment:
Could you add some `_maxAvailable` style logic to this, so if `_amountOut == type(uint256).max` then it sets `_amountOut = poolGuage.balanceOf(address(this));`

Reply:
Good call, added it!

Reply:
I've made the changes!